### PR TITLE
Add support for BLAS/LAPACK ILP64 libraries

### PR DESCRIPTION
**Pull request by dmejiar**

### CODE REVIEW

**Comment by awvwgk** [comment]
Do we need the code duplication of blas.f90 and blas64.f90 here? It would be preferable if we can just have a single file rather than two to avoid divergence.

**Comment by dmejiar** [comment]
Is not necessary. Let me remove blas64.f90 and work with macros

**Comment by awvwgk** [comment]
You could define a kind parameter `ik` and use it instead of `i8`

For the macro I would go with something along the lines of

```f90
#ifndef IK
#define IK i4
#endif

module mchrg_blas
  use mctc_env, only : sp, dp, ik => IK
  implicit none

  ! ...

end module mchrg_blas
```

**Comment by dmejiar** [comment]
Let me know if this works for you. If this is ok with you, I would also prepare PR's for `dftd4`, `s-dftd3`, and `tblite` with the same structure to enable ILP64

File: CMakeLists.txt [hunk 1]
```diff
@@ -33,6 +36,13 @@ if(NOT TARGET "OpenMP::OpenMP_Fortran" AND WITH_OpenMP)
   find_package("OpenMP" REQUIRED)
 endif()
 
+if(WITH_ILP64)
+  message(STATUS "Using LAPACK/BLAS ILP64 interface")
+  add_definitions(-DIK=i8)
+else()
+  add_definitions(-DIK=i4)
+endif()
```
Comment (awvwgk) [inline CMakeLists.txt:44]: Definitions should be added to the target and not globally

File: CMakeLists.txt [hunk 2]
```diff
@@ -16,13 +16,16 @@
 cmake_minimum_required(VERSION 3.14)
 get_directory_property(is-subproject PARENT_DIRECTORY)
 
+
+# C is required to find MKL automatically
 project(
   "multicharge"
-  LANGUAGES "Fortran"
+  LANGUAGES "Fortran" "C"
```
Comment (awvwgk) [inline CMakeLists.txt:23]: I would prefer to only depend on a Fortran compiler here.

File: config/CMakeLists.txt [hunk 3]
```diff
@@ -78,3 +81,5 @@ install(
   "${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}.pc"
   DESTINATION "${CMAKE_INSTALL_LIBDIR}/pkgconfig"
 )
+
+include(CheckFortranSourceCompiles)
```
Comment (awvwgk) [inline config/CMakeLists.txt:85]: Where is this used in the current scope?

File: config/meson.build [hunk 4]
```diff
@@ -49,15 +49,25 @@ if lapack_vendor == 'auto'
...
+  add_project_arguments('-DIK=i4', language:'fortran')
+  mkllp = ''
+  lp = ''
+endif
+
 if lapack_vendor == 'mkl'
   mkl_dep = []
   if fc_id == 'intel'
-    mkl_dep += fc.find_library('mkl_intel_lp64')
+    mkl_dep += fc.find_library('mkl_intel_'+mkllp+'lp64')
```
Comment (awvwgk) [inline config/meson.build:65]: Maybe use a ternary operator or format the string for clarity

File: src/multicharge/model.F90 [hunk 5]
```diff
@@ -14,7 +14,7 @@
 ! limitations under the License.
```
Comment (awvwgk) [inline src/multicharge/model.F90:15]: Make sure this keeps working with fpm
```suggestion

#ifndef IK
#define IK i4
#endif

```

File: src/multicharge/blas.F90 [hunk 6]
```diff
@@ -15,7 +15,7 @@
```
Comment (awvwgk) [inline src/multicharge/blas.F90:15]: Make sure this keeps working with fpm
```suggestion

#ifndef IK
#define IK i4
#endif

```

File: src/multicharge/lapack.F90 [hunk 7]
```diff
@@ -14,7 +14,7 @@
 ! limitations under the License.
```
Comment (awvwgk) [inline src/multicharge/lapack.F90:15]: Make sure this keeps working with fpm
```suggestion

#ifndef IK
#define IK i4
#endif

```

**Comment by dmejiar** [comment]
Actually, I was thinking the macros are not needed. If someone (like me) wants to use the ILP64 interface, then one can handle that with the `-Dilp64=true`  **and** giving the right fortran flags, i.e. `-Dfortran_flags="-fdefault-integer-8`.

What's your preference @awvwgk ? I can close this PR and open a new one with a clean commit history

**Comment by awvwgk** [comment]
Indeed, promoting default integers to 8 bytes (`-i8` for Intel and `-fdefault-integer-8` for GCC) and using the custom library option should make this possible without changes in the build files already.

I'm still open to officially support ilp64 with the tblite stack.

File: config/CMakeLists.txt (hunk 3 above)
Comment (dmejiar) [reply config/CMakeLists.txt]: I went back to the simpler `Findcustom-blas.cmake` and `Findcustom-lapack.cmake` and this is no longer needed

File: CMakeLists.txt (hunk 2 above)
Comment (dmejiar) [reply CMakeLists.txt]: I have removed this in my latest commit, but the automatic MKL detection with `cmake` will not longer work. See bottom of https://cmake.org/cmake/help/latest/module/FindLAPACK.html

**Comment by awvwgk** [comment]
If the C compiler is needed for MKL detection, we could add it optionally in Findcustom-blas.cmake, if the correct `BLA_VENDOR` is set.

**Comment by dmejiar** [comment]
The following works for either LP64 or ILP64 if `cmake >= 3.22` and `$MKLROOT` is set. For older versions of `cmake` the user should provide the correct `BLA_VENDOR`


```diff
diff --git a/config/cmake/Findcustom-blas.cmake b/config/cmake/Findcustom-blas.cmake
index 3dfe5c1..7f71e62 100644
--- a/config/cmake/Findcustom-blas.cmake
+++ b/config/cmake/Findcustom-blas.cmake
@@ -13,6 +13,14 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+if ((BLA_VENDOR MATCHES ^Intel) OR (DEFINED ENV{MKLROOT}))
+  enable_language("C")
+endif()
+
+if(WITH_ILP64)
+  set(BLA_SIZEOF_INTEGER 8)
+endif()
+
 if(NOT BLAS_FOUND)
   find_package("BLAS")
   if(NOT TARGET "BLAS::BLAS")
diff --git a/config/cmake/Findcustom-lapack.cmake b/config/cmake/Findcustom-lapack.cmake
index c78569d..8cd9a6a 100644
--- a/config/cmake/Findcustom-lapack.cmake
+++ b/config/cmake/Findcustom-lapack.cmake
@@ -13,6 +13,14 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+if ((BLA_VENDOR MATCHES ^Intel) OR (DEFINED ENV{MKLROOT}))
+  enable_language("C")
+endif()
+
+if(WITH_ILP64)
+  set(BLA_SIZEOF_INTEGER 8)
+endif()
+
 if(NOT LAPACK_FOUND)
   find_package("LAPACK")

```

**Comment by awvwgk** [comment]
This seems like a good solution to me.

File: CMakeLists.txt [hunk 8]
```diff
@@ -16,13 +16,16 @@
...
+
+# C is required to find MKL automatically
 project(
   "multicharge"
   LANGUAGES "Fortran"
   VERSION "0.2.0"
   DESCRIPTION "Molecular structure store for testing"
 )
 
+
```
Comment (awvwgk) [inline CMakeLists.txt:28]:
```suggestion
```

File: CMakeLists.txt [hunk 9]
```diff
@@ -83,6 +90,15 @@ if(WITH_OpenMP)
     "OpenMP::OpenMP_Fortran"
   )
 endif()
+if(WITH_ILP64)
+  target_compile_definitions(
+    "${PROJECT_NAME}-lib"
+    PUBLIC -DIK=i8)
```
Comment (awvwgk) [inline CMakeLists.txt:96]:
```suggestion
    PUBLIC -DIK=i8
  )
```

File: CMakeLists.txt [hunk 10]
```diff
@@ -83,6 +90,15 @@ if(WITH_OpenMP)
...
+else()
+  target_compile_definitions(
+    "${PROJECT_NAME}-lib"
+    PUBLIC -DIK=i4)
```
Comment (awvwgk) [inline CMakeLists.txt:100]:
```suggestion
    PUBLIC -DIK=i4
  )
```

**Review by awvwgk — APPROVED** [review]
Thanks for this patch. Looks good to me now.

**Pull request merged**